Move all the pixel computations to `Sprite::compute_pixel_space_point`.

andriyDev opened the pull request:
I merged, and then added 2 commits on top. The first is a function to compute where in the texture a sprite point is. The second is rewriting the picking backend to take advantage of this. I tested this with the `sprite_picking` example and it seems to work quite well! Note this isn't just a move: I rewrote the logic. I've taken some of the logic from `crates\bevy_sprite\src\render\mod.rs:406` and mixed in some of my own ideas. Hopefully this is more clear and also handles more cases!

I haven't dug in to figuring out SpriteImageMode yet, so that might be the next step.

PS: I goofed and originally set it to merge to your main rather than your PR.

bas-ie committed:
and -> an (#16483)

Fix stray character in docs.

atlv24 committed:
cleanup bevy_render/lib.rs (#16481)

# Objective

- Refactor

## Solution

- Refactor

## Testing

- Ran 3d_scene

---

## Migration Guide

`RenderCreation::Manual` variant fields are now wrapped in a struct
called `RenderResources`

andriyDev committed:
Create `compute_pixel_space_point` to convert from a point relative to a sprite, to the pixel point to sample from.

andriyDev committed:
Rewrite the sprite picking backend to use `Sprite::compute_pixel_space_point`.

andriyDev committed:
Merge branch 'main' into HEAD

vandie commented on crates/bevy_sprite/src/picking_backend.rs:
```diff
@@ -169,36 +153,33 @@ pub fn sprite_picking(
+                            // Check the alpha is above the cutoff.
+                            color.alpha() > cutoff
                         }
+                        SpritePickingMode::BoundingBox => true,
```
if you switch to bounding box mode, does this still work. We're not actually using the in rect check anywhere below

andriyDev replied:
We've already checked that `compute_pixel_space_point` returns Ok, which means we're in the bounds of the sprite. So we've already done the check.

andriyDev replied:
I added a comment for this though, since it definitely was confusing.

vandie commented on crates/bevy_render/src/settings.rs:
Can we try to limit this PR to the change that we're actually trying to accomplish. Including random other changes like the structure of this file will just make it harder to merge the final change

andriyDev replied:
That just comes from the merge commit. It's already on bevy main.

vandie reviewed (commented):
I think this is likely on the right track tbh

andriyDev committed:
Add a comment to clarify that we know the cursor is in the bounds of the sprite.

andriyDev committed:
Make the sprite_picking system private, but then re-export everything in picking_backend.

Previously SpriteBackendSettings was private, making it impossible to change the SpritePickingMode. Now we can!

andriyDev commented:
Ok apparently I'm being gaslit. This doesn't seem to be working correctly anymore?? I had tested it visually, but doing it now produces the wrong results. I need to investigate.

andriyDev committed:
Remove flipping the y-value of the cursor from the picking backend.

This is already taken care of in `compute_pixel_space_point`.

andriyDev commented:
Fixed the above issue. I must have broken things as I was cleaning up my commits. I also tested using the `SpritePickingMode::BoundingBox` and it looks to be working correctly.

andriyDev committed:
Rename `SpriteBackendSettings` to `SpritePickingSettings`.

This matches the `SpritePickingPlugin`.
